lib/ivis_opengl/pielighting.cpp

```diff
@@ -0,0 +1,225 @@
+	data = std::make_unique<PIELIGHT[]>(width * height);
+	ASSERT(data != nullptr, "Out of memory");
+}
+
+
+LightingData& getCurrentLightingData()
+{
+	static LightingData scene;
+	return scene;
+}
+
+LightMap& getCurrentLighmapData()
+{
+	static LightMap lightmap;
+	return lightmap;
+}
+
+namespace {
+	BoundingBox getLightBoundingBox(const LIGHT& light)
+	{
+		glm::vec3 center = light.position;
+		center.z *= -1.;
+		float range = light.range;
+		glm::vec3 horizontal(1.0, 0., 0.);
+		glm::vec3 vertical(0.0, 1.0, 0.);
+		glm::vec3 forward(0.0, 0., 1.0f);
+
+		return BoundingBox{
+			center - horizontal * range - vertical * range - forward * range,
+			center - horizontal * range - vertical * range + forward * range,
+			center - horizontal * range + vertical * range - forward * range,
+			center - horizontal * range + vertical * range + forward * range,
+			center + horizontal * range - vertical * range - forward * range,
+			center + horizontal * range - vertical * range + forward * range,
+			center + horizontal * range + vertical * range - forward * range,
+			center + horizontal * range + vertical * range + forward * range
+		};
+	}
+
+}
+
+void renderingNew::LightingManager::ComputeFrameData(const LightingData& data, LightMap&, const glm::mat4& worldViewProjectionMatrix)
+{
+	PointLightBuckets result;
+
+	// Pick the first lights inside the view frustrum
+	auto viewFrustrum = IntersectionOfHalfSpace{
+		[](glm::vec3 in) { return in.x >= -1.f; },
+		[](glm::vec3 in) { return in.x <= 1.f; },
+		[](glm::vec3 in) {
+			if (gfx_api::context::get().isYAxisInverted())
+				return -in.y >= -1.f ;
+			return in.y >= -1.f ;
+		},
+		[](glm::vec3 in) {
+			if (gfx_api::context::get().isYAxisInverted())
+				return -in.y <= 1.f;
+			return in.y <= 1.f;
+		},
+		[](glm::vec3 in) { return in.z >= 0; },
+		[](glm::vec3 in) { return in.z <= 1; }
+	};
+
+	std::vector<LIGHT> culledLights;
+	for (const auto& light : data.lights)
+	{
+		if (culledLights.size() >= gfx_api::max_lights)
+		{
+			break;
+		}
+		auto clipSpaceBoundingBox = transformBoundingBox(worldViewProjectionMatrix, getLightBoundingBox(light));
+		if (!isBBoxInClipSpace(viewFrustrum, clipSpaceBoundingBox))
+		{
+			continue;
+		}
+		culledLights.push_back(light);
+	}
+
+
+	for (size_t lightIndex = 0; lightIndex < culledLights.size(); lightIndex++)
+	{
+		const auto& light = culledLights[lightIndex];
+		result.positions[lightIndex].x = light.position.x;
+		result.positions[lightIndex].y = light.position.y;
+		result.positions[lightIndex].z = light.position.z;
+		result.colorAndEnergy[lightIndex].x = light.colour.byte.r / 255.f;
+		result.colorAndEnergy[lightIndex].y = light.colour.byte.g / 255.f;
+		result.colorAndEnergy[lightIndex].z = light.colour.byte.b / 255.f;
+		result.colorAndEnergy[lightIndex].w = light.range;
+	}
+
+	// Iterate over all buckets
+	size_t overallId = 0;
+	size_t bucketId = 0;
+	std::array<size_t, gfx_api::max_indexed_lights * 4> lightList;
```

Comment:
Why `4`? Can you please add a comment regarding this?

Reply:
@vlj

Reply:
I added the comment